resolver/type_syntax.cc

```diff
@@ -506,6 +506,25 @@ core::TypePtr interpretTCombinator(core::MutableContext ctx, ast::Send *send, co
                 e.setHeader("Only top-level T.self_type is supported");
             }
             return core::Types::untypedUntracked();
+        case core::Names::attachedClass()._id:
+            // The conditions for using T.attached_class are the same as
+            // T.self_type, so we just piggy-back off of that flag.
```

Comment:
Is this true? I thought that `T.self_type` would be restricted to be used in fewer places.

It seems that `T.attached_class` should only be restricted to places where type members could have been used previously, which to my understanding was "everywhere".

For example, If we only allow top-level `T.attached_class`:

```ruby
# typed: true
class Box
  extend T::Sig
  extend T::Generic

  Elem = type_template(upper: Box)

  sig {returns(T::Array[Elem])}
  def self.foo
    []
  end

  sig {returns(T::Array[T.self_type])}
  def self.bar
    []
  end
end
```

```

# typed: true
class Box
  extend T::Sig
  extend T::Generic

  Elem = type_template(upper: Box)

  sig {returns(T::Array[Elem])}
  def self.foo
    []
  end

  sig {returns(T::Array[T.self_type])}
  def self.bar
    []
  end
end


editor.rb:13: Only top-level T.self_type is supported https://srb.help/5004
    13 |  sig {returns(T::Array[T.self_type])}
                                ^^^^^^^^^^^
Errors: 1
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20Box%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AGeneric%0A%0A%20%20Elem%20%3D%20type_template(upper%3A%20Box)%0A%0A%20%20sig%20%7Breturns(T%3A%3AArray%5BElem%5D)%7D%0A%20%20def%20self.foo%0A%20%20%20%20%5B%5D%0A%20%20end%0A%0A%20%20sig%20%7Breturns(T%3A%3AArray%5BT.self_type%5D)%7D%0A%20%20def%20self.bar%0A%20%20%20%20%5B%5D%0A%20%20end%0Aend)

Reply:
My worry was how it would interact with subtyping, but I think you've convinced me that it's OK and necessary.